lib/documents.js

```diff
@@ -2263,12 +2264,143 @@ function writeAllDocumentsImpl(options) {
+    }
+    const multiplier = (jobState.jobOptions.concurrentRequests.multiplier &&
+        jobState.jobOptions.concurrentRequests.multiplier>0)?
+        jobState.jobOptions.concurrentRequests.multiplier:0;
+    maxRequesters = (multiplier >0) ? (maxRequesters*multiplier):maxRequesters;
+
+  }
+
+  for (let i = 0; i < maxRequesters; i++) {
+    jobState.requesterCount++;
+    onWriteAllDocs(operation,i);
+  }
+
+}
+
+function onWriteAllDocs(operation, writerId) {
+
+  const writeBatchArray = [];
+  const jobState = operation.jobState;
+  const batchSize = (jobState.jobOptions && jobState.jobOptions.batchSize &&
+      jobState.jobOptions.batchSize >0) ?
+      jobState.jobOptions.batchSize:100;
+
+  a:
+      for (let i = 0; i < batchSize; i++) {
+        const record = jobState.stream.read();
+
+        if(jobState.jobOptions && jobState.jobOptions.defaultMetadata) {
+          writeBatchArray.push(jobState.jobOptions.defaultMetadata);
+        }
+        if (record === null) {
+          switch (writeBatchArray.length) {
+            case 0: {
+              finishWriter(operation);
+              break a;
+            }
+            default: {
+              writeDocumentsImpl.call(jobState.docInstance, false, writeBatchArray)
+                  .result((out) =>{
+                    jobState.docsWrittenSuccessfully+= writeBatchArray.length;
+                      finishWriter(operation);
+                      })
+                  .catch((err) => {
+                    jobState.docsFailedToBeWritten+= writeBatchArray.length;
+                    jobState.error = err;
+                    jobState.stream.emit('error', err);
+                  });
+              break a;
+            }
+          }
+        }
+        writeBatchArray.push(record);
+      }
+
+  if (writeBatchArray.length !== 0) {
+    writeDocumentsImpl.call(jobState.docInstance, false, writeBatchArray)
+        .result((output) => {
+          onWriteAllDocs(operation,writerId);
+          jobState.docsWrittenSuccessfully+= writeBatchArray.length;
+          if(jobState.jobOptions && jobState.jobOptions.onBatchSuccess) {
+            const progressSoFar = {
+              docsWrittenSuccessfully: jobState.docsWrittenSuccessfully,
+              docsFailedToBeWritten: jobState.docsFailedToBeWritten
+            };
+            const documents = writeBatchArray;
+            jobState.jobOptions.onBatchSuccess(progressSoFar, documents);
+          }
+          })
+        .catch((err) => {
+          jobState.docsFailedToBeWritten+= writeBatchArray.length;
+          if(jobState.jobOptions && jobState.jobOptions.onBatchError) {
+            const progressSoFar = {
+              docsWrittenSuccessfully: jobState.docsWrittenSuccessfully,
+              docsFailedToBeWritten: jobState.docsFailedToBeWritten
+            };
+            const documents = writeBatchArray;
+            jobState.jobOptions.onBatchError(progressSoFar, documents, err);
```

Comment:
This is a good start for the error notification.  

As in Java, it will be necessary to inspect the return value to determine whether to retry the batch, skip the batch, or end the stream.